### PR TITLE
Update typehint code to allow primitive composite types in python 3.9+

### LINKED ISSUE #25146: [Bug]: PEP 585 generic support mishandles primitive composites
### What happened?

The support for PEP 585 generics added in #25054 missed removing the constraint on primitives in composites, leading to type hints like `tuple[str, list]` returning an error instead of being converted to `Tuple[str, List[Any]]`.

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by jrmccluskey**
Updates python typehinting code to allow primitive containers in composite type hints, allowing statements like `tuple[str, list]` to work since their typing module equivalent also worked. 

Fixes #25146 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jrmccluskey** [comment]
Run Python PreCommit

**Comment by jrmccluskey** [comment]
R: @damccorm 
CC: @kennknowles

File: sdks/python/apache_beam/typehints/typehints.py [hunk 1]
```diff
@@ -389,10 +389,12 @@ def validate_composite_type_param(type_param, error_msg_prefix):
   if sys.version_info.major == 3 and sys.version_info.minor >= 10:
     if isinstance(type_param, types.UnionType):
       is_not_type_constraint = False
-  is_forbidden_type = (
-      isinstance(type_param, type) and type_param in DISALLOWED_PRIMITIVE_TYPES)
+  if sys.version_info.major == 3 and sys.version_info.minor < 9:
```
Comment (damccorm) [inline sdks/python/apache_beam/typehints/typehints.py:392]: Could you add a comment explaining the pre/post 3.9, 3.10 logic in this block?

**Review by AnandInguva — APPROVED** [review]

File: sdks/python/apache_beam/typehints/typehints_test.py [hunk 2]
```diff
@@ -379,9 +379,10 @@ def test_getitem_params_must_be_type_or_constraint(self):
       typehints.Tuple[5, [1, 3]]
     self.assertTrue(e.exception.args[0].startswith(expected_error_prefix))
 
-    with self.assertRaises(TypeError) as e:
-      typehints.Tuple[list, dict]
-    self.assertTrue(e.exception.args[0].startswith(expected_error_prefix))
+    if sys.version_info < (3, 9):
+      with self.assertRaises(TypeError) as e:
+        typehints.Tuple[list, dict]
+      self.assertTrue(e.exception.args[0].startswith(expected_error_prefix))
```
Comment (damccorm) [inline sdks/python/apache_beam/typehints/typehints_test.py:385]: Can we add else clauses to these tests that validate that they don't throw?

**Review by damccorm — APPROVED** [review]
Just had some minor feedback

File: sdks/python/apache_beam/typehints/typehints_test.py (hunk 2 above)
Comment (jrmccluskey) [reply sdks/python/apache_beam/typehints/typehints_test.py:385]: Done

File: sdks/python/apache_beam/typehints/typehints.py (hunk 1 above)
Comment (jrmccluskey) [reply sdks/python/apache_beam/typehints/typehints.py:394]: Done

**Review by damccorm — APPROVED** [review]
LGTM, thanks!

**Pull request merged**